keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1503 @@
+```
+
+allows representing the new formula for a pods resource usage
+
+```
+Max ( Max( each InitContainerUse ) , Sum(Sidecar Containers) + Sum(each ContainerUse) ) + pod overhead
+```
+
+Even now, users still sometimes find how a pods effective resource requirements
+are calculated confusing or are just unaware of the formula.  The mitigating
+quality to this is that init container resource requests are usually lower than
+the sum of non-init container resource requests, and can be ignored by users in
+those cases.  Software that requires accurate pod resource requirement
+information (e.g. kube-scheduler, kubelet, autoscalers) don't have that luxury.
+Its too much to ask of users to perform this even more complex calculation
+simply to know the amount of free capacity they need for a given resource to
+allow a pod to schedule.
+
+#### Goals of exposing the Pod.ResourceAllocated field
+
+- Allow Kubernetes users to quickly identify the effective resource requirements
+  for a pending or scheduled pod directly via observing the pod status.
+- Allow consuming the Pod Requested Resources via metrics:
+  - Make sure kube_pod_resource_requests formula is up to date
+  - Consider exposing Pod Requirements as a Pod state metric via
+    https://github.com/kubernetes/kube-state-metrics/blob/main/docs/pod-metrics.md 
+- Provide a well documented, reusable, exported function to be used to calculate
+  the effective resource requirements for a v1.Pod struct.
+- Eliminate duplication of the pod resource requirement calculation within
+  `kubelet` and `kube-scheduler`.
+
+#### Implementation details
+
+We propose making two changes to satisfy the two primary consumers of the
+effective pod resource requests, users and other Kubernetes ecosystem software
+components.
+
+The first change is to add a field to `PodStatus` to represent the effective
+resource requirements for the pod.  The field is named `ResourcesAllocated` which
+matches the naming of the container status field from the KEP 1287. This field
+allows users to inspect running and pending pods to quickly determine their
+effective resource requirements. The field will be first populated by scheduler
+in the [updatePod](https://github.com/kubernetes/kubernetes/blob/815e651f397a6f5691efb26b0d4f3bddb034668c/pkg/scheduler/schedule_one.go#L943). The updating of this field would occur the existing
+[generateAPIPodStatus](https://github.com/kubernetes/kubernetes/blob/a668924cb60901b413abc1fe7817bc7969167064/pkg/kubelet/kubelet_pods.go#L1459) method.
+
+```
+// ResourcesAllocated is the effective resource requirements for this pod, taking into consideration init containers, sidecars, containers and pod overhead.
+// More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/  
+// +optional  
+ResourcesAllocated ResourceRequirements `json:"resources,omitempty" protobuf:"bytes,8,opt,name=resources"`
```

Comment:
```suggestion
// ResourcesAllocated is the effective resource requirements for this pod, taking into consideration init containers, sidecars, containers and pod overhead.
// More info: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/  
// +optional  
ResourcesAllocated ResourceList `json:"resourcesAllocated,omitempty" protobuf:"bytes,8,opt,name=resourcesAllocated"`
```

Maybe we need to change the json field name and type?
- name: resources -> resourcesAllocated
- type: ResourceRequirements -> ResourceList ?

Reply:
Yes, I changed the name a few times and missed the JSON tag.  I think a ResourceList makes sense as well, it also matches the container `ResourcesAllocated` field in the in-place resizing KEP.

Reply:
fixed